File: AdventureWorks/AdventureWorks.Client.Common/DataObjects/Sales/SalesOrderDetailObjectCustomized.cs
using AdventureWorks.Services;
using System;
using Xomega.Framework;

namespace AdventureWorks.Client.Objects
{
    public class SalesOrderDetailObjectCustomized : SalesOrderDetailObject
    {
        public SalesOrderDetailObjectCustomized()
        {
        }

        public SalesOrderDetailObjectCustomized(IServiceProvider serviceProvider) : base(serviceProvider)
        {
        }

        // construct properties and child objects
        protected override void Initialize()
        {
            base.Initialize();
            // add custom construction code here
        }

        // perform post intialization
        protected override void OnInitialized()
        {
            base.OnInitialized();

            ProductIdProperty.SetCascadingProperty(Enumerations.Product.Attributes.ProductSubcategoryId, SubcategoryProperty);
            ProductIdProperty.CascadingMatchNulls = true; // blank subcategory will display products with no categories
            ProductIdProperty.Change += (sender, e) =>
            {
                if (e.Change.IncludesValue() && !ProductIdProperty.IsNull())
                {
                    UnitPriceProperty.SetValue(ProductIdProperty.Value[Enumerations.Product.Attributes.ListPrice]);
                    UpdateLineTotal(sender, e);
                }
            };

            SpecialOfferIdProperty.LocalCacheLoader = new SpecialOfferProductReadListCacheLoader(ServiceProvider);
            SpecialOfferIdProperty.SetCacheLoaderParameters(Enumerations.SpecialOfferProduct.Parameters.ProductId, ProductIdProperty);
            SpecialOfferIdProperty.Change += (sender, e) =>
            {
                if (e.Change.IncludesValue() && !SpecialOfferIdProperty.IsNull())
                {
                    UnitPriceDiscountProperty.SetValue(SpecialOfferIdProperty.Value[Enumerations.SpecialOfferProduct.Attributes.Discount]);
                    UpdateLineTotal(sender, e);
                }
            };
            OrderQtyProperty.Change += UpdateLineTotal;
        }

        private void UpdateLineTotal(object sender, PropertyChangeEventArgs e)
        {
            if (e.Change.IncludesValue())
            {
                var price = UnitPriceProperty.Value ?? 0;
                var qty = OrderQtyProperty.Value ?? 0;
                var discount = UnitPriceDiscountProperty.Value ?? 0;
                LineTotalProperty.SetValue(price * (1 - discount) * qty);
            }
        }
    }
}

File: AdventureWorks/AdventureWorks.Client.Spa/DataObjects/Sales/SalesOrderDetailObjectCustomized.ts

import { SalesOrderDetailObject as GeneratedDataObject } from 'DataObjects/Sales/SalesOrderDetailObject';
import { SpecialOfferProductReadListCacheLoader } from '../../CacheLoaders/SpecialOfferProductReadListCacheLoader';
import { SpecialOfferProduct, Product } from '../../Enumerations/Enumerations';
import { Header } from 'xomega';

export class SalesOrderDetailObject extends GeneratedDataObject {

    // construct properties and child objects
    init() {
        super.init();
        // add custom construction code here
    }

    // perform post intialization
    onInitialized() {
        this.ProductId.setCascadingProperty(Product.Attributes.ProductSubcategoryId, this.Subcategory);
        this.ProductId.cascadingMatchNulls = true; // blank subcategory will display products with no categories
        this.ProductId.InternalValue.subscribe(() => {
            var hdr: Header = this.ProductId.InternalValue();
            if (hdr) this.UnitPrice.InternalValue(hdr.getAttribute(Product.Attributes.ListPrice));
            this.updateLineTotal();
        });

        this.SpecialOfferId.LocalCacheLoader = new SpecialOfferProductReadListCacheLoader();
        this.SpecialOfferId.setCacheLoaderParameters(SpecialOfferProduct.Parameters.ProductId, this.ProductId);
        this.SpecialOfferId.InternalValue.subscribe(() => {
            var hdr: Header = this.SpecialOfferId.InternalValue();
            if (hdr) this.UnitPriceDiscount.InternalValue(hdr.getAttribute(SpecialOfferProduct.Attributes.Discount));
            this.updateLineTotal();
        });

        this.OrderQty.InternalValue.subscribe(() => this.updateLineTotal());

        super.onInitialized();
    }

    updateLineTotal() {
        var price = this.UnitPrice.InternalValue() || 0;
        var qty = this.OrderQty.InternalValue() || 0;
        var discount = this.UnitPriceDiscount.InternalValue() || 0;
        this.LineTotal.InternalValue(price * (1 - discount) * qty);
    }
}